pkg/datapath/linux/ipsec/ipsec_linux.go

```diff
@@ -1121,3 +1121,32 @@ func StartStaleKeysReclaimer(ctx context.Context) {
 		}
 	}()
 }
+
+// We need to install xfrm state for the local router (cilium_host) early
+// in daemon init path. This is to ensure that we have the xfrm state in
+// place before we advertise the routerIP where other nodes may potentially
+// pick it up and start sending traffic to us. This was previously racing
+// and creating XfrmInNoState errors because other nodes picked up node
+// update before Xfrm config logic was in place. So special case init the
+// rule we need early in init flow.
+func Init() error {
+	outerLocalIP := node.GetInternalIPv4Router()
+	wildcardIP := net.ParseIP("0.0.0.0")
+	localCIDR := node.GetIPv4AllocRange().IPNet
+	localWildcardIP := &net.IPNet{IP: wildcardIP, Mask: net.IPv4Mask(0, 0, 0, 0)}
+
+	if _, err := ipSecReplaceStateIn(outerLocalIP, wildcardIP, false); err != nil {
+		return fmt.Errorf("unable to replace local state: %s", err)
+	}
+	if err := ipSecReplacePolicyIn(localWildcardIP, localCIDR, wildcardIP, outerLocalIP); err != nil {
+		if !os.IsExist(err) {
+			return fmt.Errorf("unable to replace policy in: %s", err)
+		}
+	}
+	if err := IpSecReplacePolicyFwd(localCIDR, outerLocalIP); err != nil {
+		if !os.IsExist(err) {
+			return fmt.Errorf("unable to replace policy fwd: %s", err)
+		}
+	}
```

Comment:
There's a lot of code duplication between this here and `enableIPsecIPv4`/`UpsertIPsecEndpoint`. These XFRM configs also don't exactly match those installed by `enableIPsecIPv4` in that the latter uses `n.nodeConfig.IPv4PodSubnets` as the destination CIDRs for XFRM IN policies. Finally, the IPv6 counterpart is missing.

Maybe to fix all of the above we could move the `newNode.IsLocal()` case of `enableIPsecIPv{4,6}` to this `Init` function instead of duplicating?

Reply:
One way to refactor would be to make it this?
```
func Init() error {
        outerLocalIP := node.GetInternalIPv4Router()                                                                                                                                   
        wildcardIP := net.ParseIP("0.0.0.0")                                                                                                                                                                                           
        localCIDR := node.GetIPv4AllocRange().IPNet                                                                                                                
        localWildcardIP := &net.IPNet{IP: wildcardIP, Mask: net.IPv4Mask(0, 0, 0, 0)}                                                                      

         _, err := UpsertIPsecEndpoint(localCIDR, localWildcardIP, outerLocalIP, wildcardIP, 0, IPDitIn, gsldr)
        return err;
}
```
I believe that is equivalent would probably want to double check. I prefer that to reaching into pkg/datapath/linux/node.go and start refactoring node code. I think it should probably be done but not in the context of a fix.